TLS handshake failure on 0.68.1 due to NPE

After trying out 0.68.1, which changed the way TLS handshakes work, I see this on my server. I am using TLS mutual auth in case that is affecting this case.

JRE: OpenJDK 10.0.2

```
Caused by: java.lang.NullPointerException
	at sun.net.util.IPAddressUtil.textToNumericFormatV4(IPAddressUtil.java:49) ~[?:?]
	at sun.net.util.IPAddressUtil.isIPv4LiteralAddress(IPAddressUtil.java:241) ~[?:?]
	at sun.security.util.HostnameChecker.isIpAddress(HostnameChecker.java:139) ~[?:?]
	at sun.security.util.HostnameChecker.match(HostnameChecker.java:97) ~[?:?]
	at sun.security.ssl.X509TrustManagerImpl.checkIdentity(X509TrustManagerImpl.java:481) ~[?:?]
	at sun.security.ssl.X509TrustManagerImpl.checkIdentity(X509TrustManagerImpl.java:456) ~[?:?]
	at sun.security.ssl.X509TrustManagerImpl.checkTrusted(X509TrustManagerImpl.java:296) ~[?:?]
	at sun.security.ssl.X509TrustManagerImpl.checkClientTrusted(X509TrustManagerImpl.java:139) ~[?:?]
	at io.netty.handler.ssl.ReferenceCountedOpenSslServerContext$ExtendedTrustManagerVerifyCallback.verify(ReferenceCountedOpenSslServerContext.java:232) ~[netty-handler-4.1.27.Final.jar:4.1.27.Final]
	at io.netty.handler.ssl.ReferenceCountedOpenSslContext$AbstractCertificateVerifier.verify(ReferenceCountedOpenSslContext.java:621) ~[netty-handler-4.1.27.Final.jar:4.1.27.Final]
	... 29 more
"    
```

--- a/core/src/main/java/com/linecorp/armeria/client/HttpClientPipelineConfigurator.java
+++ b/core/src/main/java/com/linecorp/armeria/client/HttpClientPipelineConfigurator.java
@@ -31,7 +31,9 @@ import java.net.SocketAddress;
 import java.util.List;
 
 import javax.annotation.Nullable;
+import javax.net.ssl.SSLEngine;
 import javax.net.ssl.SSLException;
+import javax.net.ssl.SSLParameters;
 
 import org.reactivestreams.Subscriber;
 import org.reactivestreams.Subscription;
@@ -45,7 +47,6 @@ import com.linecorp.armeria.common.HttpObject;
 import com.linecorp.armeria.common.SessionProtocol;
 import com.linecorp.armeria.common.logging.RequestLogBuilder;
 import com.linecorp.armeria.common.util.Exceptions;
-import com.linecorp.armeria.internal.ChannelUtil;
 import com.linecorp.armeria.internal.Http1ClientCodec;
 import com.linecorp.armeria.internal.Http2GoAwayListener;
 import com.linecorp.armeria.internal.ReadSuppressingHandler;
@@ -210,7 +211,7 @@ final class HttpClientPipelineConfigurator extends ChannelDuplexHandler {
         final SslHandler sslHandler = sslCtx.newHandler(ch.alloc(),
                                                         remoteAddr.getHostString(),
                                                         remoteAddr.getPort());
-        p.addLast(ChannelUtil.configureSslHandler(sslHandler));
+        p.addLast(configureSslHandler(sslHandler));
         p.addLast(TrafficLoggingHandler.CLIENT);
         p.addLast(new ChannelInboundHandlerAdapter() {
             private boolean handshakeFailed;
@@ -272,6 +273,20 @@ final class HttpClientPipelineConfigurator extends ChannelDuplexHandler {
                 ctx.close();
             }
         });
+    }
+
+    /**
+     * Configures the specified {@link SslHandler} with common settings.
+     */
+    private static SslHandler configureSslHandler(SslHandler sslHandler) {
+        // Set endpoint identification algorithm so that JDK's default X509TrustManager implementation
+        // performs host name checks. Without this, the X509TrustManager implementation will never raise
+        // a CertificateException even if the domain name or IP address mismatches.
+        final SSLEngine engine = sslHandler.engine();
+        final SSLParameters params = engine.getSSLParameters();
+        params.setEndpointIdentificationAlgorithm("HTTPS");
+        engine.setSSLParameters(params);
+        return sslHandler;
     }
 
     // refer https://http2.github.io/http2-spec/#discover-http

--- a/core/src/main/java/com/linecorp/armeria/internal/ChannelUtil.java
+++ b/core/src/main/java/com/linecorp/armeria/internal/ChannelUtil.java
@@ -20,16 +20,12 @@ import java.util.List;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.atomic.AtomicInteger;
 
-import javax.net.ssl.SSLEngine;
-import javax.net.ssl.SSLParameters;
-
 import com.google.common.collect.ImmutableList;
 
 import io.netty.channel.Channel;
 import io.netty.channel.ChannelFutureListener;
 import io.netty.channel.EventLoopGroup;
 import io.netty.channel.epoll.EpollEventLoopGroup;
-import io.netty.handler.ssl.SslHandler;
 
 public final class ChannelUtil {
 
@@ -69,20 +65,6 @@ public final class ChannelUtil {
         }
 
         return future;
-    }
-
-    /**
-     * Configures the specified {@link SslHandler} with the common settings.
-     */
-    public static SslHandler configureSslHandler(SslHandler sslHandler) {
-        // Set endpoint identification algorithm so that JDK's default X509TrustManager implementation
-        // performs host name checks. Without this, the X509TrustManager implementation will never raise
-        // a CertificateException even if the domain name or IP address mismatches.
-        final SSLEngine engine = sslHandler.engine();
-        final SSLParameters params = engine.getSSLParameters();
-        params.setEndpointIdentificationAlgorithm("HTTPS");
-        engine.setSSLParameters(params);
-        return sslHandler;
     }
 
     private ChannelUtil() {}

--- a/core/src/main/java/com/linecorp/armeria/server/HttpServerPipelineConfigurator.java
+++ b/core/src/main/java/com/linecorp/armeria/server/HttpServerPipelineConfigurator.java
@@ -37,7 +37,6 @@ import com.google.common.collect.Iterables;
 
 import com.linecorp.armeria.common.SessionProtocol;
 import com.linecorp.armeria.common.util.Exceptions;
-import com.linecorp.armeria.internal.ChannelUtil;
 import com.linecorp.armeria.internal.Http1ObjectEncoder;
 import com.linecorp.armeria.internal.Http2GoAwayListener;
 import com.linecorp.armeria.internal.ReadSuppressingHandler;
@@ -79,7 +78,6 @@ import io.netty.handler.ssl.SslContext;
 import io.netty.handler.ssl.SslHandler;
 import io.netty.util.AsciiString;
 import io.netty.util.DomainNameMapping;
-import io.netty.util.ReferenceCountUtil;
 
 /**
  * Configures Netty {@link ChannelPipeline} to serve HTTP/1 and 2 requests.
@@ -170,7 +168,7 @@ final class HttpServerPipelineConfigurator extends ChannelInitializer<Channel> {
 
     private void configureHttps(ChannelPipeline p, @Nullable ProxiedAddresses proxiedAddresses) {
         assert sslContexts != null;
-        p.addLast(new HttpsSniHandler(sslContexts));
+        p.addLast(new SniHandler(sslContexts));
         p.addLast(TrafficLoggingHandler.SERVER);
         p.addLast(new Http2OrHttpHandler(proxiedAddresses));
     }
@@ -477,33 +475,6 @@ final class HttpServerPipelineConfigurator extends ChannelInitializer<Channel> {
         private String addAfter(ChannelPipeline p, String baseName, ChannelHandler handler) {
             p.addAfter(baseName, null, handler);
             return p.context(handler).name();
-        }
-    }
-
-    /**
-     * An {@link SniHandler} that calls {@link ChannelUtil#configureSslHandler(SslHandler)}
-     * so that an {@link SslHandler} is configured correctly.
-     */
-    private static class HttpsSniHandler extends SniHandler {
-        HttpsSniHandler(DomainNameMapping<SslContext> sslContexts) {
-            super(sslContexts);
-        }
-
-        @Override
-        protected void replaceHandler(ChannelHandlerContext ctx, String hostname, SslContext sslContext) {
-            SslHandler sslHandler = null;
-            try {
-                sslHandler = sslContext.newHandler(ctx.alloc());
-                ctx.pipeline().replace(this, SslHandler.class.getName(),
-                                       ChannelUtil.configureSslHandler(sslHandler));
-                sslHandler = null;
-            } finally {
-                // Since the SslHandler was not inserted into the pipeline,
-                // the ownership of the SSLEngine was not transferred to the SslHandler.
-                if (sslHandler != null) {
-                    ReferenceCountUtil.safeRelease(sslHandler.engine());
-                }
-            }
         }
     }
 }

